teps/0060-remote-resource-resolution.md

```diff
@@ -242,207 +255,570 @@ users can leverage those shared pipelines.
+
+When the Pipelines reconciler sees a `taskRef` or `pipelineRef` with a
+`resolver` field it creates a `ResourceRequest` object. The object is
+populated using fields from the `taskRef` or `pipelineRef`.
+
+Once resolved the contents of the requested resource are base64-encoded
+and stored in-line in the `status.data` field of the `ResourceRequest`.
+Metadata is set in the `status.annotations` field. Examples of metadata
+include the commit digest of a resolved `git` resource, a bundle's
+digest, or the resolved data's `content-type`. The `ResourceRequest` is
+marked as successfully resolved. Here is a sample of a resolved
+`ResourceRequest` with some metadata fields trimmed out for brevity:
+
+```yaml
+apiVersion: resolution.tekton.dev/v1alpha1
+kind: ResourceRequest
+spec:
+  params:
+    repository_url: https://github.com/sbwsg/experimental.git
+    branch: remote-resolution
+    path: /remote-resolution/pipeline.yaml
+status:
+  data: a2luZDogUGlwZWxpbmUKYXBpVmVyc2lvbjogdGVrdG9uLmRldi92MWJldGExCm1ldGFkYXRhOgogIG5hbWU6IHAKc3BlYzoKICB0YXNrczoKICAtIG5hbWU6IGZldGNoLWZyb20tZ2l0CiAgICB0YXNrU3BlYzoKICAgICAgc3RlcHM6CiAgICAgIC0gaW1hZ2U6IGFscGluZS9naXQKICAgICAgICBzY3JpcHQ6IHwKICAgICAgICAgIGdpdCBjbG9uZSBodHRwczovL2dpdGh1Yi5jb20vdGVrdG9uY2QvcmVzdWx0cwo=
+  annotations:
+    commit: 2c9b093e94f30f588dc798cc56a2559d4da9d573
+    content-type: application/x-yaml
+  conditions:
+  - lastTransitionTime: "2021-10-14T15:38:14Z"
+    status: "True"
+    type: Succeeded
+```
+
+If a `ResourceRequest` fails resolution then it is marked as failed with
+a `Succeeded/"False"` condition. The condition will include a
+machine-readable `reason` and human-readable `message` explaining the
+failure.
+
+### 3. Create a new Tekton Resolution project
+
+Create a new project under the `tektoncd` Github org catering specifically
+to the functionality described in this TEP. The intention will be to
+take this project from alpha to beta and eventually to a stable,
+production-ready state. The project will provide:
+
+- The `ResourceRequest` CRD and a controller managing the lifecycle for
+  that CRD.
+- Code-generated clients & supporting types for `ResourceRequests`.
+- Function and struct helpers for other projects to use.
+- A common interface for resolver implementors that abstracts away the
+  underlying use of CRDs.
+
+Webhook and controller deployments for `ResourceRequest` objects will
+run in the `tekton-remote-resolution` namespace by default. The
+namespace is intentionally separate from `tekton-pipelines` to allow
+`RBAC` that isolates the remote resolution machinery.
+
+During initial development and alpha the project will build several
+resolvers: "in-cluster", "bundle", and "git". In time we may decide to
+move these to their own repo alongside community-contributed resolvers,
+if any are written. See the [Design Details](#design-details) section
+for more on specific resolvers.
+
+The `ResourceRequest` Lifecycle Controller will be responsible for the
```

Comment:
How would it be used by tektoncd/pipeline controller/machinery ? (might be documented already elsewhere right ?)

Reply:
Not sure if this answers your question exactly but the plan is for Pipelines to interact with the resolution machinery through helper libraries that the `tektoncd/resolver` project publishes. The helpers are going to hide as much of the specifics of requesting as possible, but Pipelines may need to thread through something like a `ResourceRequest` client/lister. But the client / lister would also be imported from the `tektoncd/resolver` project - not new codegen in Pipelines.

This is described a little bit above with "Function and struct helpers for other projects to use" but I'll expand on that in the Design Details section.

To summarize my thinking very quickly, the Pipelines controllers will interact with the libraries through intermediary structs rather than "raw" CRD or HTTP request objects. So, hand-waving, the reconciler code might end up looking something like:

```go
var errTaskStillResolving = errors.New("taskRef still resolving")

// resolveTaskRun wraps helpers from the tektoncd/resolver project to implement async
// taskRef resolution.
func (r *reconciler) resolveTask(ctx context.Context, ref *v1beta1.TaskRef) (*v1beta1.Task, error) {
  if alphaFlagEnabled(ctx) {
    task := &v1beta1.Task{}
    // resp here is a struct like resolutionhelpers.Resolved{ Data: []byte, Annotations: map[string]string } or similar.
    if resp, err := resolutionhelpers.Resolve(ref.Resolver, ref.Resource); err != nil {
      if errors.Is(err, &resolutionhelpers.ErrResolutionInProgress{}) {
        return nil, errTaskStillResolving // non-fatal: a resource request has just been created or still in progress.
      } else {
        return nil, controller.NewPermanentError(err)
      }
    } else {
      if err := resolutionhelpers.Unmarshal(resp, task); err != nil {
        return nil, controller.NewPermanentError(err)
      }
      task.SetDefaults(ctx) // + do any other task initialization stuff
      return task, nil
    }
  }
}
```

Reply:
I've updated the design details section with this info.